subprojects/core/src/main/java/org/gradle/internal/exceptions/StyledException.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.exceptions;
+
+import org.apache.commons.lang.StringUtils;
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.logging.text.StyledTextOutput;
+
+import javax.annotation.Nullable;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * <p>A styled exception is an exception which can be rendered to the console
+ * with styled text (basically colors). The only requirement is to put tags
+ * between the styled elements. The tag names correspond to a {@link StyledTextOutput.Style}
+ * name.</p>
+ * <p>For example: This is an &lt;Failure&gt;error&lt;/Failure&gt;</p>
+ * <p>It's worth noting that {@link #getMessage()} would return a non-rendered text,
+ * so that tooling using the exception messages are not concerned with rendering issues.
+ * </p>
+ */
+public class StyledException extends GradleException {
+    private final static Logger LOGGER = Logging.getLogger(StyledException.class);
+
+    // Keep only the outermost style
+    private final static Pattern STYLES_REGEX = Pattern.compile("<([A-Za-z]+)>(?:<([A-Za-z]+)>)*(.*?)(?:</\\2>)*</\\1>");
+    private final String styledMessage;
+
+    public StyledException() {
+        super();
+        styledMessage = "";
+    }
+
+    public StyledException(String message) {
+        super(unstyled(message));
+        styledMessage = message;
+    }
+
+    private static String unstyled(String message) {
+        Matcher matcher = STYLES_REGEX.matcher(message);
+        StringBuffer buf = new StringBuffer();
+        while (matcher.find()) {
+            matcher.appendReplacement(buf, "$3");
+        }
+        matcher.appendTail(buf);
+        return buf.toString();
+    }
+
+    public StyledException(String message, @Nullable Throwable cause) {
+        super(unstyled(message), cause);
+        styledMessage = message;
+    }
+
+    public static String style(StyledTextOutput.Style style, String text) {
```

Comment:
I wonder if this method should be in the `StyledTextOutput.Style` (non static) so that you can directly call it on a Style. I think that makes it easier to discover. And it seems to be useful everywhere where a style is rendered.

Reply:
It's only used in the exception renderer, though, because this exception renderer uses some kind of markup.

Reply:
Thanks for explaining. I missed that this is new. I thought it was borrowed from something wie already had before. Then we can keep it like this.